File: Leetcode 1469 Find All The Lonely Nodes [easy] [tree]/Find All The Lonely Nodes [dfs].js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var getLonelyNodes = function(root) {
    let lonelyNodes = [];
    if (root === null) {
        return lonelyNodes;
    }
    
    findLonelyNodes(root, lonelyNodes);
    return lonelyNodes;
};

let findLonelyNodes = (root, lonelyNodes) => {
    if (root === null) {
        return;
    }
    
    if (root.left !== null && root.right === null) {
        lonelyNodes.push(root.left.val);
    }
    
    if (root.left === null && root.right !== null) {
        lonelyNodes.push(root.right.val);
    }
    
    if (root.left !== null) {
        findLonelyNodes(root.left, lonelyNodes);
    }
    
    if (root.right !== null) {
        findLonelyNodes(root.right, lonelyNodes);
    }
}

File: Leetcode 1469 Find All The Lonely Nodes [easy] [tree]/Find All The Lonely Nodes [dfs].java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> getLonelyNodes(TreeNode root) {
        List<Integer> lonelyNodes = new ArrayList<>();
        
        if (root == null) {
            return lonelyNodes;
        }
        
        findLonelyNodes(root, lonelyNodes);
        
        return lonelyNodes;
    }
    
    private void findLonelyNodes(TreeNode root, List<Integer> lonelyNodes) {
        if (root == null) {
            return;
        }
        
        if (root.left != null && root.right == null) {
            lonelyNodes.add(root.left.val);
        }
        
        if (root.left == null && root.right != null) {
            lonelyNodes.add(root.right.val);
        }
        
        if (root.left != null) {
            findLonelyNodes(root.left, lonelyNodes);
        }
        
        if (root.right != null) {
            findLonelyNodes(root.right, lonelyNodes);
        }
    }
}